<a href="https://colab.research.google.com/github/Annette-1/Ecuaciones_Diferenciales_Parciales/blob/main/Prueba_Rachas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <span style="color:blue;">**Prueba de Rachas(Independencia)**</span>
 <span style="color:purple;"></span>

<span style="color:magenta;">*Nombre: Jerónimo López Annette Michelle*</span>

<span style="color:brown;"></span>**Objetivo:** Realizar la prueba de independecias(Rachas) y compruebe si el ejercicio cumple con la aleatoridad de los números

<span style="color:purple;">Prueba de independencia (Prueba de Rachas)</span>
Para está prueba se considera la siguiente hipótesis nula:

$H_0:$ Los números de la muestra $\{x_i\}$ son independientes

Luego

$H_1:$ Los números de la muestra $\{x_i\}$ no son independientes

In [78]:
import numpy as np
from scipy.stats import norm

In [79]:
#Datos de la tabla 3.1 de los numeros pseudoaleatorios
# Datos originales
data = [
0.78961, 0.05230, 0.10699, 0.55877, 0.14151,
0.76086, 0.12079, 0.27738, 0.65726, 0.79269,
0.80548, 0.82654, 0.29453, 0.20852, 0.42989,
0.58518, 0.98611, 0.34488, 0.34358, 0.11537,
0.89898, 0.57880, 0.67621, 0.05010, 0.00121,
0.28269, 0.73059, 0.70119, 0.18284, 0.49962,
0.38618, 0.76910, 0.68334, 0.55170, 0.10850,
0.79882, 0.45679, 0.21631, 0.87616, 0.55743,
0.58962, 0.33216, 0.03185, 0.61168, 0.09264,
0.69623, 0.17028, 0.05475, 0.91512, 0.76262,
0.29931, 0.30861, 0.83358, 0.51781, 0.03272,
0.57410, 0.26593, 0.85903, 0.43308, 0.35286,
0.24000, 0.65559, 0.38507, 0.90829, 0.94187,
0.93655, 0.88809, 0.81772, 0.36982, 0.19904,
0.54325, 0.62400, 0.09133, 0.41678, 0.33954,
0.58244, 0.85853, 0.86752, 0.38729, 0.15506,
0.23949, 0.53559, 0.33381, 0.49883, 0.75103,
0.19962, 0.65002, 0.74579, 0.79113, 0.63453,
0.19147, 0.40644, 0.08128, 0.73435, 0.22724,
0.22287, 0.07281, 0.64183, 0.44267, 0.72102
]

<span style="color:purple;">**Paso 1**</span> Determine la secuencia $S$ de $0$ y $1$ o (+,-)

Así
 Si $$ x_i≤x_{i-1}$$ escribir 0 (o -)

Si $$ x_i>x_{i-1}$$ escribir 1 (o -
+)


La secuencia $S$ contiene $n-1$ números

In [80]:
# Función para generar la secuencia de 0 y 1
def generar_secuencia(datos):
    secuencia = np.zeros(len(datos) - 1, dtype=int)
    for i in range(1, len(datos)):
        if datos[i] > datos[i - 1]:
            secuencia[i - 1] = 1
        else:
            secuencia[i - 1] = 0
    return secuencia

<span style="color:purple;">**Paso 2**</span> Contar el número de rachas observadas $C_o$, una racha viene dada por el número de ceros y unos consecutivos que la forman

In [81]:
# Función para contar el número de corridas
def contar_corridas(secuencia):
    corridas = 1  # La primera corrida siempre comienza en 1
    for i in range(1, len(secuencia)):
        if secuencia[i] != secuencia[i - 1]:  # Si cambia de 0 a 1 o viceversa, es una nueva corrida
            corridas += 1
    return corridas

Transponemos los datos para leer por columnas

In [82]:
# Convertir a matriz de 20 filas x 5 columnas
matriz_original = np.array(data).reshape(20, 5)

# Leer por columnas usando ravel con order='F' (column-major)
datos_por_columnas = matriz_original.ravel(order='F')


In [83]:
# Generar la secuencia binaria con los datos leídos por columnas
secuencia_binaria = generar_secuencia(datos_por_columnas)

secuencia_binaria= generar_secuencia(datos_por_columnas)
print("Secuencia binaria generada:", secuencia_binaria)

Secuencia binaria generada: [0 1 0 1 0 1 1 0 1 0 1 0 1 0 1 0 0 0 1 0 1 1 1 0 1 1 0 0 0 1 0 1 1 0 1 0 1
 0 0 1 1 1 1 1 1 0 0 0 1 1 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 1 1 0 1 0 0 1 0 1
 0 1 1 0 0 0 1 0 0 0 1 0 1 0 1 0 1 1 0 1 0 1 0 0 1]


<span style="color:purple;">**Paso 3**</span> Calcular el valor esperado, y la varianza del número de rachas, así cómo el estadístico $z_0$ mediante:
$$μ_{Co}= \frac{2n-1}{3}$$
$$σ^2_{Co}=\frac{16n-29}{90}$$
 Y el estadístico de prueba es:
 $$Z_0=\frac{C_{o}-μ_{Co}}{σ^2_{Co}}$$

In [84]:
# Calcular el número de corridas
num_corridas = contar_corridas(secuencia_binaria)

# Calcular el valor esperado y la varianza de las corridas
n = len(datos_por_columnas)
valor_esperado = (2 * n - 1) / 3
varianza = (16 * n - 29) / 90
desviacion = np.sqrt(varianza)

# Calcular el estadístico Z
Z = (num_corridas - valor_esperado) / desviacion

Verificamos la prueba de hipotesis dado el estadistico para prueba de rachas:
$$Z_0=\frac{C_{o}-μ_{Co}}{σ^2_{Co}}$$


Dado que $C_o$ sigue una distribución normal de media $μ_{c_o}$ y varianza $σ^2_{C_o}$ y $Z_0$ una distribución normal (0,1). El intervalo de aceptación:
$$
-≤Z_0\le Z_{\frac{α}{2}} $$
o bien
$$|Z_0| ≤Z_{\frac{α}{2}}$$

In [85]:
# Prueba de hipótesis
alpha = 0.05
z_critico = norm.ppf(1 - alpha / 2)
intervalo_critico = [-z_critico, z_critico]
es_aleatorio = intervalo_critico[0] <= Z <= intervalo_critico[1]

In [86]:
# Resultados
print("\n" + "=" * 60)
print("RESULTADOS")
print("=" * 60)
print("Número de corridas:", num_corridas)
print("Valor esperado de corridas:", valor_esperado)
print("Varianza:", varianza)
print("Desviación estándar:", desviacion)
print("Estadístico Z:", Z)
print("Intervalo crítico (α=0.05):", intervalo_critico)
print("¿La secuencia es aleatoria?:", es_aleatorio)


RESULTADOS
Número de corridas: 70
Valor esperado de corridas: 66.33333333333333
Varianza: 17.455555555555556
Desviación estándar: 4.1779846284489315
Estadístico Z: 0.8776161218256837
Intervalo crítico (α=0.05): [np.float64(-1.959963984540054), np.float64(1.959963984540054)]
¿La secuencia es aleatoria?: True


**Interpretación:**

El estadístico $Z$ calculado (-0.0827) cae dentro del intervalo crítico $[-1.96, 1.96]$, lo que indica que no hay evidencia suficiente para rechazar la hipótesis de aleatoriedad.

<span style="color:purple;">Por lo tanto no se rechaza la hipotesis
$H_0:$ Los números de la muestra $\{x_i\}$ son independientes(aleatorios)</span>

Nota: Este ejercicio paso las pruebas de: Kolmogorov-Smirov, Rachas, Bondad de ajuste.


## <span style="color:purple;">**Conclusión**:</span>

Con un nivel de significancia del 5%, se concluye que la secuencia de números pseudoaleatorios, cuando se lee por columnas, presenta un comportamiento aleatorio. El número de corridas observado (70) es muy cercano al valor esperado teórico (66.33), lo que sugiere que no existe un patrón sistemático en la generación de los números.
Los datos pueden considerarse aptos para su uso en simulaciones y métodos probabilísticos que requieran independencia y aleatoriedad en las observaciones.